session.go

```diff
@@ -479,10 +480,23 @@ func (s *sessionImpl) StartTunnel(ctx context.Context, cfg config.Tunnel) (Tunne
 		return nil, errStartTunnel{err}
 	}
 
-	return &tunnelImpl{
+	t := &tunnelImpl{
 		Sess:   s,
 		Tunnel: tunnel,
-	}, nil
+	}
+
+	if httpServerCfg, ok := cfg.(interface {
+		HTTPServer() *http.Server
+	}); ok {
+		if srv := httpServerCfg.HTTPServer(); srv != nil {
+			srv.BaseContext = func(_ net.Listener) context.Context { return ctx }
```

Comment:
I think this might be the first time we're actually using the context passed to StartTunnel for anything *real*. The underlying libraries don't actually use it, so it's just been a future-proofing thing so far.

As a result, we haven't really discussed its semantics. Is it *just* for the Bind rpc call? Is it tied to the lifetime of the tunnel itself, and is canceling it more or less equivalent to calling Close on the tunnel? This change implies the later, but I feel like the former might make more sense.

Reply:
you are 100% right. the context should be for the lifetime of the RPC call and not used here. if we had a context for the lifetime of the tunnel itself that makes sense, but in the absence of that it seems like we should just let you use WithHTTPServer and manage that context yourself if you want to supply it

Reply:
fixed